circuits/src/parsing/serialization.rs

```diff
@@ -0,0 +1,351 @@
+    /// the converted bytes.
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, String>;
+
+    /// Deserializes a the content of a slice, and returns the result, panicking
+    /// with the corresponding error message in case the `Result` of
+    /// `self.deserialize` is an `Error`.
+    fn deserialize_unwrap(serialization: &[u8]) -> Self {
+        let mut buf = serialization;
+        match Self::deserialize(&mut buf) {
+            Ok(x) => x,
+            Err(msg) => panic!(
+                "Deserialization error. Error message:\n=====\n{}\n=====",
+                msg
+            ),
+        }
+    }
+}
+
+macro_rules! ensure_buf_len {
+    ($buf:expr, $required:expr, $type:ty) => {
+        if $buf.len() < $required {
+            return Err(format!(
+                "Buffer underflow: expected at least {} bytes for deserializing `{}`, only got {}.\n>> Buffer:\n{:?}",
+                $required,
+                stringify!($type),
+                $buf.len(),
+                $buf,
+            ));
+        }
+    };
+}
+
+impl Serialize for u8 {
+    #[cfg(test)]
+    fn serialize(&self, buf: &mut Vec<u8>) {
+        buf.push(*self)
+    }
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, String> {
+        ensure_buf_len!(buf, 1, Self);
+        let res = buf[0];
+        *buf = &buf[1..];
+        Ok(res)
+    }
+}
+
+impl Serialize for u64 {
+    #[cfg(test)]
+    fn serialize(&self, buf: &mut Vec<u8>) {
+        buf.extend(&self.to_le_bytes())
+    }
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, String> {
+        const U64_BYTES: usize = u64::BITS as usize / 8;
+        ensure_buf_len!(buf, U64_BYTES, Self);
+        let res = Self::from_le_bytes(buf[..U64_BYTES].try_into().unwrap());
+        *buf = &buf[U64_BYTES..];
+        Ok(res)
+    }
+}
+
+// `usize` is serialized as `u64` to avoid the issue where the serializing and
+// deserializing machines have different architectures.
+impl Serialize for usize {
+    #[cfg(test)]
+    fn serialize(&self, buf: &mut Vec<u8>) {
+        (*self as u64).serialize(buf)
+    }
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, String> {
+        let val = u64::deserialize(buf)?;
+        if val > (usize::MAX as u64) {
+            Err(format!(
+                "Cannot deserialize value {} into usize on this platform",
+                val
+            ))
+        } else {
+            Ok(val as usize)
+        }
+    }
+}
+
+// It could be possible to optimise the serialization of booleans by encoding
+// packs of at most 8 into a single byte. However, in the case of the
+// serialization of the `RawAutomaton` type, this would only save two bytes per
+// serialized automaton, which is quite insignificant.
+impl Serialize for bool {
+    #[cfg(test)]
+    fn serialize(&self, buf: &mut Vec<u8>) {
+        if *self { 1u8 } else { 0u8 }.serialize(buf)
+    }
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, String> {
+        Ok(u8::deserialize(buf)? == 1)
+    }
```

Comment:
any value other than 1 will return false. Is that expected? If yes, you could also avoid implementing Serialize for `bool`, and treating them as `u8`s.